src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -12,99 +13,322 @@ namespace System.Net.Sockets
-                if (err != Interop.Error.SUCCESS)
+                //
+                // Create the event port and buffer
+                //
+                if (Interop.Sys.CreateSocketEventPort(out _port) != Interop.Error.SUCCESS)
+                {
+                    throw new InternalException();
+                }
+                if (Interop.Sys.CreateSocketEventBuffer(EventBufferCount, out _buffer) != Interop.Error.SUCCESS)
                 {
                     throw new InternalException();
                 }
 
-                // The native shim is responsible for ensuring this condition.
-                Debug.Assert(numEvents > 0);
+                //
+                // Create the pipe for signalling shutdown, and register for "read" events for the pipe.  Now writing
+                // to the pipe will send an event to the event loop.
+                //
+                int* pipeFds = stackalloc int[2];
+                if (Interop.Sys.Pipe(pipeFds, Interop.Sys.PipeFlags.O_CLOEXEC) != 0)
+                {
+                    throw new InternalException();
+                }
+                _shutdownReadPipe = pipeFds[Interop.Sys.ReadEndOfPipe];
+                _shutdownWritePipe = pipeFds[Interop.Sys.WriteEndOfPipe];
 
-                for (int i = 0; i < numEvents; i++)
+                if (Interop.Sys.DangerousTryChangeSocketEventRegistration(_port, _shutdownReadPipe, Interop.Sys.SocketEvents.None, Interop.Sys.SocketEvents.Read, ShutdownHandle) != Interop.Error.SUCCESS)
                 {
-                    var handle = (GCHandle)(IntPtr)_buffer[i].Data;
-                    var context = (SocketAsyncContext)handle.Target;
+                    throw new InternalException();
+                }
 
-                    if (context != null)
+                //
+                // Start the event loop on its own thread.
+                //
+                Task.Factory.StartNew(
+                    EventLoop,
+                    CancellationToken.None,
+                    TaskCreationOptions.LongRunning,
+                    TaskScheduler.Default);
+            }
+            catch
+            {
+                if (_shutdownReadPipe != -1)
+                {
+                    Interop.Sys.Close((IntPtr)_shutdownReadPipe);
+                }
+                if (_shutdownWritePipe != -1)
+                {
+                    Interop.Sys.Close((IntPtr)_shutdownWritePipe);
+                }
+                if (_buffer != null)
+                {
+                    Interop.Sys.FreeSocketEventBuffer(_buffer);
+                }
+                if (_port != -1)
+                {
+                    Interop.Sys.CloseSocketEventPort(_port);
+                }
+                throw;
+            }
+        }
+
+        private void EventLoop()
+        {
+            try
+            {
+                bool shutdown = false;
+                while (!shutdown)
```

Comment:
What happens if shutdown is requested and there are more than EventBufferCount events to be processed when we get a shutdown request?  I guess that can't happen because you only signal for shutdown when there are 0 outstanding handles?